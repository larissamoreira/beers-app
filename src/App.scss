@use './variables';

* {
  margin: 0;
}

html {
  box-sizing: border-box;
}
*,
*:before,
*:after {
  box-sizing: inherit;
}

body {
  display: grid;
  place-items: center;
  background-color: $backgroundColor;
  height: 100vh;
  width: 100vw;
}