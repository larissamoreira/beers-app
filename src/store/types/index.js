export const GET_BEERS = 'GET_BEERS'
export const GET_BEERS_SUCCESS = 'GET_BEERS_SUCCESS'
export const GET_BEERS_FAILURE = 'GET_BEERS_FAILURE'
