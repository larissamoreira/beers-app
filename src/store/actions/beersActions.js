import { GET_BEERS, GET_BEERS_SUCCESS, GET_BEERS_FAILURE } from "../types/index";
import { getBeerList } from '../../services/api';

export const getBeers = () => ({
  type: GET_BEERS,
})

export const getBeersSuccess = (beers) => ({
  type: GET_BEERS_SUCCESS,
  payload: beers,
})

export const getBeersFailure = () => ({
  type: GET_BEERS_FAILURE,
})

export function fetchBeers(page = 1) {
  return async (dispatch) => {
    try {
      const data = await getBeerList(page);
      dispatch(getBeersSuccess(data));
    } catch (error) {
      dispatch(getBeersFailure())
    }
  }
}
