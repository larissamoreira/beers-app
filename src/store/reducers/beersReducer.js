import { GET_BEERS, GET_BEERS_SUCCESS, GET_BEERS_FAILURE } from "../types/index";

export const initialState = {
  beers: [],
  loading: false,
  hasErrors: false,
}

export default function beersReducer(state = initialState, action) {
  switch (action.type) {
    case GET_BEERS:
      return { ...state, loading: true }
    case GET_BEERS_SUCCESS:
      return { beers: action.payload, loading: false, hasErrors: false }
    case GET_BEERS_FAILURE:
      return { ...state, loading: false, hasErrors: true }
    default:
      return state
  }
}
