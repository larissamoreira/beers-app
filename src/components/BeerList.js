import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom';

import { fetchBeers } from '../store/actions/beersActions'

const BeerList = ({ dispatch, loading, beers, hasErrors, next = 1 }) => {
  useEffect(() => {
    dispatch(fetchBeers(next))
  }, [dispatch])

  const renderBeers = () => {
    if (loading) return <p>Loading...</p>
    if (hasErrors) return <p>Unable to display beer list.</p>
    return beers.map((beer) => <Link to={`/beer/${beer.id}`} key={beer.id}><p>{beer.name} - {beer.tagline}</p></Link>)
  }

  return (
    <section>
      <h1>Beers</h1>
      {beers && renderBeers()}
      {next && <button onClick={() => dispatch(fetchBeers(next))}>Go to page {next}</button>}
    </section>
  )
}

const mapStateToProps = (state) => ({
  loading: state.beers.loading,
  beers: state.beers.beers,
  hasErrors: state.beers.hasErrors,
  next: state.beers.next
})

export default connect(mapStateToProps)(BeerList)
