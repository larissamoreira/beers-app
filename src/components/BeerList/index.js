import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom';

import { fetchBeers } from '../../store/actions/beersActions'
import './style.scss'

const BeerList = ({ dispatch, loading, beers, hasErrors, next = 1 }) => {
  useEffect(() => {
    dispatch(fetchBeers(next))
  }, [dispatch])

  const renderBeers = () => {
    if (loading) return <p>Loading...</p>
    if (hasErrors) return <p>Unable to display beer list.</p>
    return (
      <ul className="beersList--list">
        {beers.map((beer) =>
          <li
            className="beersList--item"
            key={beer.id}>
            <Link
              to={`/beer/${beer.id}`}
              className="beersList--item--link">
              <p><strong>{beer.name}</strong> - {beer.tagline}</p>
            </Link>
          </li>
        )}
      </ul>
    )
  }

  return (
    <section className="beersList--container">
      <h1 className="beersList--title">Beers</h1>
      {beers && renderBeers()}
      {next &&
        <button
          onClick={() => dispatch(fetchBeers(next))}
          className="button">Go to page {next}
        </button>
      }
    </section>
  )
}

const mapStateToProps = (state) => ({
  loading: state.beers.loading,
  beers: state.beers.beers,
  hasErrors: state.beers.hasErrors,
  next: state.beers.next
})

export default connect(mapStateToProps)(BeerList)
