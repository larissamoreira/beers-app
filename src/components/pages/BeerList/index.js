import React, { useEffect, useRef } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom';

import { fetchBeers } from '../../../store/actions/beersActions'
import './style.scss'

const BeerList = ({ dispatch, loading, beers, hasErrors, next = 1 }) => {
  const firstRender = useRef(true);

  useEffect(() => {
    if (firstRender.current) {
      dispatch(fetchBeers(next))
      firstRender.current = false;
    }
    handleScroll()
  }, [])

  const handleScroll = () => {
    if (window.innerHeight + document.documentElement.scrollTop + 10 >= document.documentElement.offsetHeight) {
      if (!next || hasErrors) return;

      if (!loading) dispatch(fetchBeers(next))
    }
  };

  window.onscroll = handleScroll;

  const renderBeers = () => {
    if (loading) return <p>Loading...</p>
    if (hasErrors) return <p>Unable to display beer list.</p>
    return (
      <ul className="beersList--list">
        {Object.keys(beers).map((beer) =>
          <li
            className="beersList--item"
            key={beers[beer].id}>
            <Link
              to={`/beer/${beers[beer].id}/`}
              className="beersList--item--link"
              id={beers[beer].id}>
              <p><strong>{beers[beer].name}</strong> - {beers[beer].tagline}</p>
            </Link>
          </li>
        )}
      </ul>
    )
  }

  return (
    <section className="beersList--container">
      <h1 className="beersList--title"><i className="em em-beers"></i> Beers</h1>
      {beers && renderBeers()}
    </section>
  )
}

const mapStateToProps = (state) => ({
  loading: state.beers.loading,
  beers: state.beers.beers,
  hasErrors: state.beers.hasErrors,
  next: state.beers.next
})

export default connect(mapStateToProps)(BeerList)
