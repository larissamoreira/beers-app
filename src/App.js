import React from 'react'
import {
  BrowserRouter as Router,
  Route,
  Switch,
} from 'react-router-dom'
import { Provider } from 'react-redux'
import store from './store/index'

import BeerList from './components/BeerList'
import BeerDetail from './components/BeerDetail'
import './App.scss'

const App = () => {
  return (
    <Provider store={store}>
      <Router>
        <Switch>
          <Route path="/" exact>
            <BeerList />
          </Route>
          <Route path="/beer/:id">
            <BeerDetail />
          </Route>
        </Switch>
      </Router>
    </Provider>
  )
}

export default App
