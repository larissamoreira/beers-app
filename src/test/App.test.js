import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import App from '../App'
import BeerList from '../components/pages/BeerList';

import thunk from 'redux-thunk'
import { fetchBeers } from '../store/actions/beersActions';
import beersReducer from '../store/reducers/beersReducer';
import { GET_BEERS_SUCCESS } from "../store/types/index";

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

const initialState = {
  beers: {},
  loading: false,
  hasErrors: false,
  next: 1
}

const beer = {
  name: 'Buzz',
  id: 1,
  tagline: 'A Real Bitter Experience.',
  description: 'A light, crisp and bitter IPA brewed with English and American hops. A small batch brewed only once.',
  image_url: 'https://images.punkapi.com/v2/keg.png'
}

describe('App', () => {
  test('should render App component', () => {
    render(<App />, { wrapper: MemoryRouter });
    expect(screen.getByText(/Beers/i)).toBeInTheDocument();
  });
});

describe('BeersList', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      beers: {},
      loading: false,
      hasErrors: false,
      next: 1
    });

    store.dispatch = jest.fn();

    renderer.act(() => {
      component = renderer.create(
        <Provider store={store}>
          <BeerList />
        </Provider>
      );
    })
  });

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
})

describe('Test thunk action', () => {
  it('should execute fetchPosts thunk successfully', () => {
    const store = mockStore(initialState);

    const expectedType = "GET_BEERS_SUCCESS"

    const expectedPayload = beer

    return store.dispatch(fetchBeers())
      .then(() => {
        const actualAction = store.getActions()
        expect(actualAction[0].payload[1]).toEqual(expectedPayload);
        expect(actualAction[0].type).toEqual(expectedType);
      });
  })
});

describe('Test reducer', () => {
  it('should change state successfully', () => {
    const action = {
      type: GET_BEERS_SUCCESS,
      payload: [beer]
    }
    const nextState = beersReducer(initialState, action);
    expect(nextState).toEqual({
      beers: {
        '0': {
          ...beer
        }
      },
      loading: false,
      hasErrors: false,
      next: 2
    });
  });
});
