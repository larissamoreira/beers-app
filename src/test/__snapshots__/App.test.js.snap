// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeersList should render with given state from Redux store 1`] = `
<section
  className="beersList--container"
>
  <h1
    className="beersList--title"
  >
    <i
      className="em em-beers"
    />
     Beers
  </h1>
</section>
`;
